### PR TITLE
Added support for Synchronous CLI

### PR DESCRIPTION
**Pull request by chinmay-browserstack**
Support for Synchronous CLI
Added `sync` parameter in options using which CLI will wait for snapshot/comparison to be completed and will return a response with details of snapshot/comparison

### CODE REVIEW

**Review by Amit3200 — DISMISSED** [review]

File: packages/core/src/api.js [hunk 1]
```diff
@@ -91,14 +95,36 @@ export function createPercyServer(percy, port) {
...
-      return res.json(200, { success: true });
+      let data;
+      if (req.body.sync) percy.syncMode = true;
+      if (percy.syncMode) {
+        try {
+          const id = await promiseWrapper(percy.snapshot, req.body);
+          data = await percy.client.getSnapshotDetails(id);
+        } catch (e) {
+          data = { message: e.message };
+        }
```
Comment (Amit3200) [inline packages/core/src/api.js:106]: Move to a function and use that.

File: packages/core/src/wait-for-snapshot.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    if (interval < MIN_POLLING_INTERVAL) {
+      interval = MIN_POLLING_INTERVAL;
+    }
+
+    this.timer = setTimeout(async () => {
+      const snapshotIds = this.snapshots.map(snap => snap.id);
+      const response = await this.percy.client.getStatus(this.type, snapshotIds);
+
+      // Poll atleast once in 2 min
+      let nextPoll = 120;
```
Comment (Amit3200) [inline packages/core/src/wait-for-snapshot.js:39]: move to a constants variable.

File: packages/core/src/wait-for-snapshot.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      this.run(optimalNextPollTime * 1000);
+    }, interval);
+  }
+
+  // If there are other snapshots which can be completed in next
+  // 5 seconds, calling after x seconds will reduce network call
+  getOptimalPollTime(lowestPollTime, now) {
+    let pollTime = lowestPollTime;
+    this.snapshots.forEach((snap) => {
+      const snapPollTime = snap.nextPoll - now;
```
Comment (Amit3200) [inline packages/core/src/wait-for-snapshot.js:77]: use full names only don't use `snap` for `snapshot`

File: packages/core/src/wait-for-snapshot.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    return pollTime;
+  }
+
+  stop() {
+    this.exit = true;
+    if (this.timer) clearTimeout(this.timer);
+  }
+}
+
+export class SnapshotData {
```
Comment (Amit3200) [inline packages/core/src/wait-for-snapshot.js:91]: Add getResult of snapshot here.

File: packages/core/src/wait-for-snapshot.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  constructor(id, time, resolve, reject) {
```
Comment (Amit3200) [inline packages/core/src/wait-for-snapshot.js:92]: may be make it hash.
`id, time, {resolve, reject}`

File: packages/cli-snapshot/src/snapshot.js [hunk 6]
```diff
@@ -83,9 +83,11 @@ export const snapshot = command('snapshot', {
       let config = { sitemap, include, exclude };
       options = merge(percy.config.sitemap, config);
     }
+    const snapshotPromise = {};
 
     yield* percy.yield.start();
-    yield* percy.yield.snapshot(options);
+    yield* percy.yield.snapshot(options, snapshotPromise);
+    yield* await handleSyncSnapshot(options, snapshotPromise, percy);
```
Comment (ninadbstack) [inline packages/cli-snapshot/src/snapshot.js:90]: do we need this ? whats usecase of sync in snapshot command ?

File: packages/core/src/wait-for-snapshot.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+      if (snaphotPollTime - lowestPollTime <= THRESHOLD_OPTIMAL_POLL_TIME) {
+        pollTime = Math.max(pollTime, snaphotPollTime);
+      }
+    });
+    return pollTime;
+  }
+
+  stop() {
+    this.exit = true;
+    if (this.timer) clearTimeout(this.timer);
```
Comment (ninadbstack) [inline packages/core/src/wait-for-snapshot.js:85]: we should also reject all pending requests

File: packages/core/src/wait-for-snapshot.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  run(interval = MIN_POLLING_INTERVAL) {
+    if (this.exit) return;
+    this.running = true;
+
+    if (interval < MIN_POLLING_INTERVAL) {
+      interval = MIN_POLLING_INTERVAL;
+    }
+
+    this.log.debug(`Polling for snapshot status in ${interval}ms`);
+    this.timer = setTimeout(async () => {
```
Comment (ninadbstack) [inline packages/core/src/wait-for-snapshot.js:37]: its easier to manage lifecycle with a loop than setTimeout

Comment (ninadbstack) [inline packages/core/src/wait-for-snapshot.js:37]: its easier to manage a loop with sleep than timeout

File: packages/core/src/wait-for-snapshot.js (hunk 4 above)
Comment (chinmay-browserstack) [reply packages/core/src/wait-for-snapshot.js:92]: Created a separate function for results

**Review by shantanuk-browserstack — APPROVED** [review]

File: packages/core/src/api.js [hunk 9]
```diff
@@ -108,7 +120,7 @@ export function createPercyServer(percy, port) {
         }, { snake: true }))
       ].join('');
 
-      const response = { success: true };
+      const response = { success: true, data };
```
Comment (Amit3200) [inline packages/core/src/api.js:123]: confirm this once that app-percy is fine and rest looks good to me.

**Review by Amit3200 — APPROVED** [review]
Approved with one comment.

File: packages/client/src/client.js [hunk 10]
```diff
@@ -182,6 +182,19 @@ export class PercyClient {
     return this.get(`builds/${buildId}`);
   }
 
+  async getSnapshotDetails(snapshotId, type = 'snapshot') {
+    if (!['snapshot', 'comparison'].includes(type)) throw new Error('Invalid type passed');
+    validateId(type, snapshotId);
+    return this.get(`${type}s/${snapshotId}?sync=true`);
```
Comment (ninadbstack) [inline packages/client/src/client.js:188]: Fix naming / create 2 functions

File: packages/client/test/client.test.js [hunk 11]
```diff
@@ -293,6 +293,39 @@ describe('PercyClient', () => {
...
+      await expectAsync(client.getStatus('snap', [1]))
+        .toBeRejectedWithError('Invalid type passed');
+    });
+
+    it('gets snapshot data', async () => {
+      api.reply('/job_status?sync=true&type=snapshot&id=1,2', () => [200, { data: '<<status-data-snapshot>>' }]);
+      api.reply('/job_status?sync=true&type=comparison&id=1,2', () => [200, { data: '<<status-data-comparison>>' }]);
+      await expectAsync(client.getStatus('snapshot', [1, 2])).toBeResolvedTo({ data: '<<status-data-snapshot>>' });
+      await expectAsync(client.getStatus('comparison', [1, 2])).toBeResolvedTo({ data: '<<status-data-comparison>>' });
+    });
```
Comment (ninadbstack) [inline packages/client/test/client.test.js:326]: split specs

File: packages/core/src/wait-for-snapshot.js [hunk 12]
```diff
@@ -0,0 +1,97 @@
+import logger from '@percy/logger';
+
+const MIN_POLLING_INTERVAL = 5_000;
+// Poll atleast once in 2 min
+const MAX_POLLING_INTERVAL = 120; // in seconds
```
Comment (ninadbstack) [inline packages/core/src/wait-for-snapshot.js:5]: change one name to make same ms/s as others

File: packages/core/src/wait-for-snapshot.js [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+const THRESHOLD_OPTIMAL_POLL_TIME = 5;
+
+export class WaitForSnapshot {
```
Comment (ninadbstack) [inline packages/core/src/wait-for-snapshot.js:8]: [Optional]
```suggestion
export class WaitForResult {
```

File: packages/core/src/snapshot.js [hunk 14]
```diff
@@ -205,6 +206,17 @@ export function validateSnapshotOptions(options) {
   return { clientInfo, environmentInfo, ...migrated };
 }
 
+export async function handleSyncSnapshot(snapshotPromise, percy, type) {
+  let data;
+  try {
+    const snapshotId = await snapshotPromise;
+    data = await percy.client.getSnapshotDetails(snapshotId, type);
+  } catch (e) {
+    data = { message: e.message };
```
Comment (ninadbstack) [inline packages/core/src/snapshot.js:215]: error key

File: packages/core/src/percy.js [hunk 15]
```diff
@@ -365,6 +380,26 @@ export class Percy {
     if (this.testing && JSON.stringify(this.testing) === JSON.stringify({})) { return true; }
     return tokenType !== 'web';
   }
+
+  syncMode(options) {
+    let syncMode = false;
+    if (this.config?.snapshot?.sync) syncMode = true;
+    if (options?.sync) syncMode = true;
```
Comment (ninadbstack) [inline packages/core/src/percy.js:387]: condition for sync explicitly set to false when global config is true

File: packages/core/test/api.test.js [hunk 16]
```diff
@@ -151,7 +152,29 @@ describe('API Server', () => {
     });
 
     expect(percy.snapshot).toHaveBeenCalledOnceWith(
-      { 'test-me': true, me_too: true }
+      { 'test-me': true, me_too: true }, {}
+    );
```
Comment (ninadbstack) [inline packages/core/test/api.test.js:156]: specifically add expectation that sync function is not called.

File: packages/core/test/snapshot.test.js [hunk 17]
```diff
@@ -451,6 +566,25 @@ describe('Snapshot', () => {
...
+      domSnapshot: testDOM,
+      sync: true
+    }, promise);
+    await percy.idle();
+    const data = await handleSyncSnapshot(promise['test snapshot'], percy, 'snapshot');
+    expect(data).toEqual(api.DEFAULT_REPLIES['/snapshots/4567?sync=true']()[1]);
+    expect(logger.stderr).toEqual([]);
+    expect(logger.stdout).toEqual([
+      '[percy] Snapshot taken: test snapshot',
+      '[percy] Waiting for snapshot test snapshot to be completed'
```
Comment (ninadbstack) [inline packages/core/test/snapshot.test.js:584]:
```suggestion
      '[percy] Waiting for snapshot `test snapshot` to be completed'
```

File: packages/core/test/wait-for-snapshot.test.js [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+    expect(waitForSnapshot.exit).toEqual(true);
+    expect(logger.stderr).toEqual([]);
+  });
+
+  it('should handle stop when run is in progress', async () => {
+    waitForSnapshot.push(snapshot);
+    waitForSnapshot.stop();
+    await jasmine.clock().tick(5000);
+    expect(getStatusMock).not.toHaveBeenCalled();
+    expect(mockReject).toHaveBeenCalledOnceWith(new Error('Unable to process synchronous results as the CLI was exited while awaiting completion of the snapshot.'));
```
Comment (ninadbstack) [inline packages/core/test/wait-for-snapshot.test.js:137]: timer expetations

File: packages/core/src/wait-for-job.js [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+      if (jobPollTime - lowestPollTime <= THRESHOLD_OPTIMAL_POLL_TIME) {
+        pollTime = Math.max(pollTime, jobPollTime);
+      }
+    });
+    return pollTime;
+  }
+
+  stop() {
+    this.exit = true;
+    if (this.timer) clearTimeout(this.timer);
```
Comment (ninadbstack) [inline packages/core/src/wait-for-job.js:83]:
```suggestion
    if (this.timer) {
       clearTimeout(this.timer);
       this.timer = null;
     }
```

File: packages/core/src/wait-for-job.js [hunk 20]
```diff
@@ -0,0 +1,97 @@
+import logger from '@percy/logger';
+
+const MIN_POLLING_INTERVAL = 5_000;
+// Poll atleast once in 2 min
+const MAX_POLLING_INTERVAL_SECONDS = 120;
+const THRESHOLD_OPTIMAL_POLL_TIME = 5;
+
+export class WaitForJob {
```
Comment (ninadbstack) [inline packages/core/src/wait-for-job.js:8]: // will be either comparison or snapshot job

**Review by ninadbstack — APPROVED** [review]
LGTM

File: packages/core/src/percy.js [hunk 21]
```diff
@@ -365,6 +380,27 @@ export class Percy {
...
+    if (options?.sync === false) syncMode = false;
```
Comment (Amit3200) [inline packages/core/src/percy.js:388]: This condition is breaking global config basically overrides if sync = true is there in global config and nothing is passed in screenshot command.

Comment (chinmay-browserstack) [reply packages/core/src/percy.js:388]: SDK issue. fixed in SDK

**Pull request merged**